icu4c/source/i18n/listformatter.cpp

```diff
@@ -35,34 +36,193 @@
+                    SimpleFormatter thenTwo, SimpleFormatter elseTwo,
+                    SimpleFormatter thenEnd, SimpleFormatter elseEnd) :
+      PatternHandler(elseTwo, elseEnd),
+      test(testFunc),
+      thenTwoPattern(thenTwo),
+      thenEndPattern(thenEnd) { }
+
+  virtual ~ContextualHandler();
+
+  virtual PatternHandler* clone() const {
+      return new ContextualHandler(
+          test, thenTwoPattern, twoPattern, thenEndPattern, endPattern);
+  }
+
+  virtual const SimpleFormatter& getTwoPattern(
+      const UnicodeString& text) const {
+      return (test)(text) ? thenTwoPattern : twoPattern;
+  }
+
+  virtual const SimpleFormatter& getEndPattern(
+      const UnicodeString& text) const {
+      return (test)(text) ? thenEndPattern : endPattern;
+  }
+
+ private:
+  bool (*test)(const UnicodeString&);
+  SimpleFormatter thenTwoPattern;
+  SimpleFormatter thenEndPattern;
+};
+
+ContextualHandler::~ContextualHandler() {
+}
+
+static const char16_t *spanishY = u"{0} y {1}";
+static const char16_t *spanishE = u"{0} e {1}";
+static const char16_t *spanishO = u"{0} o {1}";
+static const char16_t *spanishU = u"{0} u {1}";
+static const char16_t *hebrewVav = u"{0} \u05D5{1}";
+static const char16_t *hebrewVavDash = u"{0} \u05D5-{1}";
+
+// Condiction to change to e.
+// Starts with "hi" or "i" but not with "hie" nor "hia"
+static bool shouldChangeToE(const UnicodeString& text) {
+    int32_t len = text.length();
+    if (len == 0) return false;
+    // Case insensitive match hi but not hie nor hia.
+    if ((text[0] == 'h' || text[0] == 'H') &&
+        ((len > 1) && (text[1] == 'i' || text[1] == 'I')) &&
+        ((len == 2) || !(text[2] == 'a' || text[2] == 'A' || text[2] == 'e' || text[2] == 'E')))
+        return true;
+    // Case insensitive for "start with i"
+    if (text[0] == 'i' || text[0] == 'I') return true;
+    return false;
+}
+
+// Condiction to change to u.
+// Starts with "o", "ho", and "8". Also "11" by itself.
+// re: ^((o|ho|8).*|11)$
+static bool shouldChangeToU(const UnicodeString& text) {
+    int32_t len = text.length();
+    if (len == 0) return false;
+    // Case insensitive match o.* and 8.*
+    if (text[0] == 'o' || text[0] == 'O' || text[0] == '8') return true;
+    // Case insensitive match ho.*
+    if ((text[0] == 'h' || text[0] == 'H') &&
+        ((len > 1) && (text[1] == 'o' || text[1] == 'O'))) return true;
+    // match "^11$" and "^11 .*"
+    if ((len >= 2) && text[0] == '1' && text[1] == '1' && (len == 2 || text[2] == ' ')) return true;
+    return false;
+}
+
+// Condiction to change to VAV follow by a dash.
+// Starts with non Hebrew letter.
+static bool shouldChangeToVavDash(const UnicodeString& text) {
+  if (text.length() == 0) return false;
+  UErrorCode status = U_ZERO_ERROR;
+  return uscript_getScript(text.char32At(0), &status) != USCRIPT_HEBREW;
+}
+
+PatternHandler* createPatternHandler(
+    const char* lang, const UnicodeString& two, const UnicodeString& end,
```

Comment:
Please indent parameters by another 4 so we see the step to the body indentation.